### PR TITLE
Specify all `__init__` params for MSN and GNN

### LINKED ISSUE #74: List all estimator parameters for grid search compatability
I was testing out `GridSearchCV` with `GNNRegressor` and realized that we have an issue with our `__init__` signatures for `GNNRegressor` and `MSNRegressor`. Currently those only explicitly accept `n_components` and take all other parameters (like `n_neighbors`) via `**kwargs`, but grid searching requires all searchable parameters to be explicitly stated to satisfy this check:

https://github.com/scikit-learn/scikit-learn/blob/24b581bce88598d92954182fab53e760cf92e6bd/sklearn/base.py#L277C13-L282C18

I think this should be an easy fix of just explicitly adding all the `KNeighborsRegressor.__init__` parameters, which should also make for a nicer docstring and better Intellisense. We should probably also add a quick test for grid search functionality while we're at it.

`EuclideanKNNRegressor` and `MalahanobisKNNRegressor` should be okay because they don't override their `__init__` methods.

### PR DESCRIPTION
**Pull request by aazuspan**
This would close #74 by explicitly specifying all the inherited `__init__` parameters for `MSNRegressor` and `GNNRegressor` via the new `OrdinationKNeighborsRegressor` abstract base class, so that they can be used in a grid search. I also added a basic test to run `GridSearchCV` on all estimators.

@grovduck, while I was adding the init params, I figured I might as well write up docstrings for those estimators as well. I copied directly from `KNeighborsRegressor` for all the shared parameters and attributes, but had to wing it a little bit for the specifics. Could you give those a look and edit as needed? Also, should we add some references or any more implementation info?

~I also considered making a shared mixin class to avoid the duplicated `__init__` method - something like `ComponentReducerRegressorMixin` - but wanted to keep it simple for the first pass and get your thoughts. Worth abstracting that out? I didn't want to cause any confusion with the existing `ComponentReducerMixin` that's used for the transformers.~ Ended up doing this with `OrdinationKNeighborsRegressor`.

### CODE REVIEW

File: src/sknnr/_gnn.py [hunk 1]
```diff
@@ -1,12 +1,94 @@
...
+    See `sklearn.neighbors.KNeighborsRegressor` for more information on parameters
+    and implementation.
+
+    Parameters
+    ----------
+    n_neighbors : int, default=5
+        Number of neighbors to use by default for `kneighbors` queries.
+    n_components : int, default=None
+        Number of components to keep during CCA transformation. If None, all
+        components are kept. If n_components is less than the number of available
```
Comment (grovduck) [inline src/sknnr/_gnn.py:29]:
```suggestion
        components are kept. If n_components is greater than the number of available
```

File: src/sknnr/_msn.py [hunk 2]
```diff
@@ -1,12 +1,94 @@
...
+    See `sklearn.neighbors.KNeighborsRegressor` for more information on parameters
+    and implementation.
+
+    Parameters
+    ----------
+    n_neighbors : int, default=5
+        Number of neighbors to use by default for `kneighbors` queries.
+    n_components : int, default=None
+        Number of components to keep during CCorA transformation. If None, all
+        components are kept. If n_components is less than the number of available
```
Comment (grovduck) [inline src/sknnr/_msn.py:29]:
```suggestion
        components are kept. If n_components is greater than the number of available
```

**Review by grovduck — COMMENTED** [review]
All looks great and thanks so much for busting out these docstrings!  I have a few minor edits/suggestions in there but nothing big.

**Comment by grovduck** [comment]
> Also, should we add some references or any more implementation info?

It's probably a good idea to add in references for these.  These would be the same as in the `README.md` file:

- Moeur M, Stage AR. 1995. Most Similar Neighbor: An Improved Sampling Inference Procedure for Natural Resources Planning. Forest Science, 41(2), 337–359.
- Ohmann JL, Gregory MJ. 2002. Predictive Mapping of Forest Composition and Structure with Direct Gradient Analysis and Nearest Neighbor Imputation in Coastal Oregon, USA. Canadian Journal of Forest Research, 32, 725–741.

I don't think there is any additional implementation details that need to be added.

> I also considered making a shared mixin class to avoid the duplicated `__init__` method - something like `ComponentReducerRegressorMixin` - but wanted to keep it simple for the first pass and get your thoughts. Worth abstracting that out? I didn't want to cause any confusion with the existing `ComponentReducerMixin` that's used for the transformers.

This is a great question as it would be nice to remove the duplication for the two `__init__` methods.  I think that as long as we can keep it straight in our minds, having a mixin for these two classes makes sense.  I know what you're saying about using `ComponentReducerRegressorMixin` (quite a mouthful!) and the possibility of it being confused with `ComponentReducerMixin`.  What do you think about a new base name entirely: something like `OrdinationRegressorMixin`?  Would that be too confusing?  Ordination usually implies dimension reduction as well, but this is adding a new term.  However, it's unlikely that users would ever be using this mixin class directly.

File: src/sknnr/_gnn.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (aazuspan) [reply src/sknnr/_gnn.py:29]: Oops, good catch!

File: src/sknnr/_gnn.py [hunk 3]
```diff
@@ -1,12 +1,94 @@
...
+        Leaf size passed to BallTree or KDTree.
+    p : int, default=2
+        Power parameter for the Minkowski metric.
+    metric : str or callable, default='minkowski'
+        The distance metric to use for the tree, calculated in CCA space.
+    metric_params : dict, default=None
+        Additional keyword arguments for the metric function.
+    n_jobs : int, default=None
+        The number of parallel jobs to run for neighbors search. None means 1 unless
+        in a joblib.parallel_backend context. -1 means using all processors.
```
[the remark this answers is not in the document]
Comment (aazuspan) [reply src/sknnr/_gnn.py:45]: Yes, good call! I missed back-ticks in a few others spots too. I *think* the double back-tick is just for RST/Sphinx - I don't see any difference between single and double back-ticks in the `mkdocs` output.

File: src/sknnr/_gnn.py [hunk 4]
```diff
@@ -1,12 +1,94 @@
...
 
 class GNNRegressor(YFitMixin, TransformedKNeighborsRegressor):
-    def __init__(self, n_components=None, **kwargs):
-        super().__init__(**kwargs)
+    """
+    Regression using Gradient Nearest Neighbor (GNN) imputation.
+
+    The target is predicted by local interpolation of the targets associated with
+    the nearest neighbors in the training set, calculated in transformed Canonical
+    Correspondence Analysis (CCA) space.
```
[the remark this answers is not in the document]
Comment (aazuspan) [reply src/sknnr/_gnn.py:18]: I like it!

File: src/sknnr/_gnn.py [hunk 5]
```diff
@@ -1,12 +1,94 @@
...
+
+    See `sklearn.neighbors.KNeighborsRegressor` for more information on parameters
+    and implementation.
```
[the remark this answers is not in the document]
Comment (aazuspan) [reply src/sknnr/_gnn.py:21]: Do you think it's good as-is, or should we lean more heavily on referencing `KNeighborsRegressor`? Currently I just omitted some specific details like how each of the `algorithm` options works, but otherwise this largely duplicates the docstring.

[the remark this answers is not in the document]
Comment (grovduck) [reply src/sknnr/_gnn.py:21]: My feeling is that it is good as-is,

File: src/sknnr/_gnn.py [hunk 6]
```diff
@@ -1,12 +1,94 @@
...
+
+    Parameters
+    ----------
+    n_neighbors : int, default=5
+        Number of neighbors to use by default for `kneighbors` queries.
+    n_components : int, default=None
```
[the remark this answers is not in the document]
Comment (aazuspan) [reply src/sknnr/_gnn.py:27]: Great question that I wasn't sure on either. [`numpydoc`](https://numpydoc.readthedocs.io/en/latest/format.html#parameters) doesn't seem to include `None` as a parameter type in these cases, even though it would be needed in the type annotation itself. That probably makes sense from a readability perspective, since `default=None` implies that `None` is supported.  

While I was looking into that, I noticed that `numpydoc` *does* distinguish between `default=None` and `optional`, and suggests the latter when `None` signifies an undefined parameter, e.g.:

```
n_components : int, optional
```

When `None` is directly used in the function, they suggest specifying `default=None`, e.g.:

```
metric_params : dict, default=None
```

Based on that, it looks like `n_components` and `n_jobs` should definitely be marked `optional`, whereas `metric_params` probably makes sense as `default=None`. I'll make those changes.

**Comment by aazuspan** [comment]
> It's probably a good idea to add in references for these. These would be the same as in the README.md file:

Good call, added!

> What do you think about a new base name entirely: something like OrdinationRegressorMixin? Would that be too confusing? Ordination usually implies dimension reduction as well, but this is adding a new term. 

I think I like this suggestion. Just a couple additional thoughts:

1. Correct me if I'm wrong, but I assume ordination also implies that the transformer could be fit with separate species data `y_fit`? If so, does it make sense to fold the `YFitMixin` into this new base class?
3. Because this base class defines an `__init__` with all the `KNeighborsRegressor` params and `n_components`, it's going to be inherently tied to kNN. Should we make that clear in the name, e.g. `OrdinationKNNMixin` or `OrdinationKNeighborsMixin`?

File: src/sknnr/_gnn.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (grovduck) [reply src/sknnr/_gnn.py]: That's great.  Thanks for digging into `numpydoc`.

**Comment by grovduck** [comment]
> 1. Correct me if I'm wrong, but I assume ordination also implies that the transformer could be fit with separate species data `y_fit`? If so, does it make sense to fold the `YFitMixin` into this new base class?

I might be misclassifying, but I think PCA is an example of an ordination method with dimensionality reduction that only relies on the `X` variables, so I don't think we want to include `YFitMixin` with the new base class.  I think somewhere along the line, we thought we could easily create a "PCA-NN" method that would calculate a PCA transformation with a `KNeighborsRegressor` imputation, and that could inherit from this base class.

> 2\. Because this base class defines an `__init__` with all the `KNeighborsRegressor` params and `n_components`, it's going to be inherently tied to kNN. Should we make that clear in the name, e.g. `OrdinationKNNMixin` or `OrdinationKNeighborsMixin`?

I like your latter suggestion.  `OrdinationKNeighborsMixin` sounds great to me!

**Comment by aazuspan** [comment]
> I might be misclassifying, but I think PCA is an example of an ordination method with dimensionality reduction that only relies on the X variables, so I don't think we want to include YFitMixin with the new base class. I think somewhere along the line, we thought we could easily create a "PCA-NN" method that would calculate a PCA transformation with a KNeighborsRegressor imputation, and that could inherit from this base class.

Ah, I'd forgotten that! Definitely best to keep `YFitMixin` separate.

> OrdinationKNeighborsMixin sounds great to me!

While I was implementing this, I ended up making [a few changes](https://github.com/lemma-osu/sknnr/pull/75/commits/4c5cc2d99d603d2f3db841c512b6d34182c8ad66) to try and simplify the inheritance of MSN and GNN and get type checking passing. Instead of treating the new base class as a mixin with no superclass (which caused `mypy` issues due to the `__init__` signature), I'm currently treating it as a new extended subclass of `TransformedKNeighborsRegressor` called `OrdinationKNeighborsRegressor`, and having `MSNRegressor` and `GNNRegressor` inherit from that instead.

I think that overall simplifies things and should give us a little better type checking, but wanted to confirm that it makes sense to you as well, and I'm not overlooking any drawbacks.

File: src/sknnr/_msn.py [hunk 7]
```diff
@@ -24,46 +24,50 @@ class MSNRegressor(YFitMixin, TransformedKNeighborsRegressor):
...
     algorithm : {'auto', 'ball_tree', 'kd_tree', 'brute'}, default='auto'
         Algorithm used to compute the nearest neighbors.
     leaf_size : int, default=30
-        Leaf size passed to BallTree or KDTree.
+        Leaf size passed to `BallTree` or `KDTree`.
     p : int, default=2
         Power parameter for the Minkowski metric.
     metric : str or callable, default='minkowski'
-        The distance metric to use for the tree, calculated in CCorA space.
+        The distance metric to use for the tree, calculated in CCA space.
```
Comment (grovduck) [inline src/sknnr/_msn.py:35]: Ooh, just caught that this should be `CCorA` space.  It looks like you changed it in this commit.  Was there a reason for that?

Comment (aazuspan) [reply src/sknnr/_msn.py]: Over-eager copy pasting, thanks for catching it!

**Comment by aazuspan** [comment]
Thanks for the review @grovduck!
